source/kernel/src/forth.rs

```diff
@@ -191,6 +299,147 @@ impl Params {
+        BagOfHolding {
+            idx: 0,
+            inner,
+            kernel,
+        }
+    }
+
+    /// Generate a new, currently unused, Bag of Holding token.
+    ///
+    /// This token will never be zero, and a given bag of holding will never
+    /// contain two items with the same token.
+    fn next_idx(&mut self) -> i32 {
+        // todo we could do better lol
+        loop {
+            self.idx = self.idx.wrapping_add(1);
+            if self.idx == 0 {
+                continue;
+            }
+            if !self.inner.iter().any(|(idx, _)| *idx == self.idx) {
+                return self.idx;
+            }
+        }
+    }
+
+    /// Place an item into the bag of holding
+    ///
+    /// The item will be allocated into a `HeapBox`, and a non-zero i32 token will
+    /// be returned on success.
+    ///
+    /// Returns an error if the Bag of Holding is full.
+    ///
+    /// At the moment there is no way to "unregister" an item, it will exist until
+    /// the BagOfHolding is dropped.
+    pub async fn register<T>(&mut self, item: T) -> Result<i32, ()>
+    where
+        T: 'static,
+    {
+        if self.inner.is_full() {
+            return Err(());
         }
+        let leaked = self.kernel.heap().allocate(item).await.leak().cast::<()>();
+        let idx = self.next_idx();
+        let tid = TypeId::of::<T>();
+
+        // Should never fail - we checked whether we are full above already
+        self.inner
+            .push((
+                idx,
+                BohValue {
+                    tid,
+                    leaked,
+                    dropfn: dropfn::<T>,
+                },
+            ))
+            .map_err(drop)?;
+        Ok(idx)
+    }
+
+    /// Attempt to retrieve an item from the Bag of Holding
+    ///
+    /// This will only succeed if the same `T` is used as was used when calling
+    /// [`BagOfHolding::register()`], and if the token matches the one returned
+    /// by `register()`.
+    ///
+    /// If the token is unknown, or the `T` does not match, `None` will be returned
+    ///
+    /// At the moment, no `get_mut` functionality is provided. It *could* be, as the
+    /// Bag of Holding represents ownership of the contained items, however it would
+    /// not be possible to retrieve multiple mutable items at the same time. This
+    /// could be added in the future if desired.
+    pub fn get<T>(&self, idx: i32) -> Option<&T>
+    where
+        T: 'static,
+    {
+        let val = &self.inner.iter().find(|(i, _item)| *i == idx)?.1;
+        let tid = TypeId::of::<T>();
+        if val.tid != tid {
+            return None;
+        }
+        let t = val.leaked.cast::<T>();
+        unsafe { Some(t.as_ref()) }
+    }
+}
+
+/// A container item for a type-erased object
+struct BohValue {
+    /// The type id of the `T` pointed to by `leaked`
+    tid: TypeId,
+    /// A non-null pointer to a `T`, contained in a leaked `HeapBox<T>`.
+    leaked: NonNull<()>,
```

Comment:
nit, take it or leave it: i don't love the name `leaked` for this, since it's...not actually leaked, really. i might just call it `value` or something. up to you though.